src/QsCompiler/CompilerOptimizations/PureCircuitAPI.fs

```diff
@@ -0,0 +1,211 @@
+﻿module Microsoft.Quantum.QsCompiler.CompilerOptimization.PureCircuitAPI
+
+open System.Collections.Immutable
+open System.Numerics
+open Microsoft.Quantum.QsCompiler.DataTypes
+open Microsoft.Quantum.QsCompiler.SyntaxTokens
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+
+open ComputationExpressions
+open Types
+open Utils
+open Printer
+
+
+/// An identifier for a qubit or for an array of qubits
+type QubitReference = Qubit of int | QubitArray of int
+
+/// Any constant expression
+type Const =
+| UnitValue
+| ValueTuple            of Const list
+| IntLiteral            of int64
+| BigIntLiteral         of BigInteger
+| DoubleLiteral         of double
+| BoolLiteral           of bool
+| StringLiteral         of string
+| ResultLiteral         of QsResult
+| PauliLiteral          of QsPauli
+| RangeLiteral          of Const * Const
+| ValueArray            of Const list
+| ArrayItem             of Const * Const
+| QubitReference        of QubitReference
+
+/// A call to a quantum gate, with the given functors and arguments
+type GateCall = {
+    gate:     QsQualifiedName
+    adjoint:  bool
+    controls: Const list
+    arg:      Const
+}
+
+/// A pure, parameter-free quantum circuit
+type Circuit = {
+    numQubitRefs: int
+    gates: GateCall list
+}
+
+/// Metadata used for constructing circuits
+type private CircuitContext = {
+    qubitReferences: string list
+}
+
+
+/// Converts a TypedExpression to a Const, mutating the given CircuitContext as needed.
+/// Returns None if the given expression cannot be converted to a constant.
+let rec private exprToConst (cc: CircuitContext ref) (expr: TypedExpression): Const option =
```

Comment:
What is the difference between this routine here and the constant folding one? Is there a way to leverage the same logic?

Reply:
No, the code is almost entirely different, and I don't think I could reuse the logic